### PR TITLE
Add example to try configuring capsense (touch) with raw PAC access

### PR DESCRIPTION
**Pull request by fudanchii**
translated partially from [chopstx csn driver](https://github.com/im-tomu/chopstx/blob/efm32/u2f/csn.c)

Example for configuring capsense with efm32's peripheral reflect system.

~still in progress~

### CODE REVIEW

**Comment by lucab** [comment]
Nice!
I think we are going in parallel down a similar path. I tried starting with a simpler capsense-to-led-loop instead, which seems to work. But I'm a very puzzled as it seems to be ok on channel0, but not on channel1 (i.e. the left/red-side button).

My (unpolished) code is:
<details>
 <summary>capsense.rs</summary>

```rust
#![no_std]
#![no_main]

extern crate panic_halt;
use cortex_m_rt::entry;
use tomu::{prelude::*, Tomu};

#[entry]
fn main() -> ! {
    let mut tomu = Tomu::take().unwrap();

    // Route clock to ACMP0.
    tomu.CMU.hfperclken0.write(|w| w.acmp0().set_bit());

    /* Usual boring LEDs+delay+watchdog setup. */

    let clk_mgmt = tomu.CMU.constrain().split();
    let gpio = tomu.GPIO.split(clk_mgmt.gpio).pins();
    let mut leds = led::LEDs::new(gpio.pa0.into(), gpio.pb7.into());
    let mut delay = systick::SystickDelay::new(tomu.SYST.constrain(), clk_mgmt.hfcoreclk);
    tomu.watchdog.disable();
    leds.green.off();
    leds.red.off();

    /* Experimental ACMP logic (unpolished) */

    tomu.ACMP0.ctrl.write(|w| w.hystsel().hyst1());
    tomu.ACMP0.inputsel.write(|w| w.csressel().res1());
    tomu.ACMP0.inputsel.write(|w| w.csresen().set_bit());
    tomu.ACMP0.inputsel.write(|w| w.negsel().capsense());
    tomu.ACMP0.ctrl.write(|w| w.en().set_bit());

    // A large delay for ACMP warmup.
    delay.delay_ms(600u16);

    // XXX(lucab): ch1 does not seem to work?!
    loop {
        // Select input channel 0.
        tomu.ACMP0.inputsel.write(|w| w.possel().ch0());

        // Read ACMP output and drive the led accordingly.
        if tomu.ACMP0.status.read().acmpout().bit_is_set() {
            leds.red.off();
        } else {
            leds.red.on();
        }
        delay.delay_ms(10u16);
    }
}
```
</details>

Does using `.ch1()` instead work for you (either on mine or your code)? Am I maybe missing something?

**Comment by fudanchii** [comment]
wow, you beat me to it, yes it seems the ch1 doesn't work in tomu board, if I'm not mistaken even all the C examples for tomu's capsense isn't work with ch1. This is probably h/w related, but I can't confirm.

**Comment by lucab** [comment]
Bummer :(
After some searches, I found https://github.com/im-tomu/tomu-samples/issues/33 too.

**Comment by fudanchii** [comment]
My implementation currently stuck at the interrupt handling, it seems either I mistakenly configure the interrupts or the irq/vectors config returned by toboot is wrong.

Haven't had the chance to debug this.

@lucab If you'd like do you want to push your code as an examples for touch button?
I will need to ensure the interrupt is working first for my implementation :sweat:

**Comment by lucab** [comment]
Sure thing.

As a different topic, I actually also wanted to start playing with timers and interrupts. Possibly with a simple "timer plus led-toggling" example. Do you have a known-working subset of this PR timer logic, or is it completely broken? Does one of either timer0 or timer1 maybe work at least, like the buttons? Or is the problem generically with interrupt routing?

**Comment by lucab** [comment]
@fudanchii in case it could be helpful for timerX too, I pushed a working example with led toggling based on RTC interrupts at https://github.com/fudanchii/imtomu-rs/pull/32.

**Comment by fudanchii** [comment]
yes, timer works, thanks for the pointer :sparkles: 

<details>
<summary>timer1</summary>

```rust
#![no_std]
#![no_main]

extern crate panic_halt;

use core::ops::DerefMut;

use core::cell::{RefCell};

use cortex_m_rt::entry;
use cortex_m::{asm, interrupt as intr, interrupt::Mutex};

use efm32::interrupt;
use efm32_hal::gpio::*;
use tomu::{prelude::*, Tomu};

static ITIMER1: Mutex<RefCell<Option<efm32::TIMER1>>> = Mutex::new(RefCell::new(None));
static GREENLED: Mutex<RefCell<Option<led::LED<pins::PA0<Output<OpenDrain<Normal, PullUp>>>>>>> = Mutex::new(RefCell::new(None));

#[entry]
fn main() -> ! {
    let mut tomu = Tomu::take().unwrap();

    tomu.watchdog.disable();

    tomu.CMU.hfperclken0.write(|w| w.timer1().set_bit());


    // constrain CMU and split into device clocks
    // so we can enable gpio with its owned clock
    let clk_mgmt = tomu.CMU.constrain().split();
    let gpio = tomu.GPIO.split(clk_mgmt.gpio).pins();

    // create tomu's led instance from gpio pin
    let leds = led::LEDs::new(gpio.pa0.into(), gpio.pb7.into());

    let mut red = leds.red;
    let green = leds.green;

    red.off();

    intr::free(|lock| GREENLED.borrow(lock).replace(Some(green)));

    tomu.TIMER1.ctrl.write(|w| w.presc().div1024());

    //21MHz / 1024 = ~20508 (1 second)
    tomu.TIMER1.top.write(|w| unsafe { w.top().bits(20508u16) });
    tomu.TIMER1.ien.write(|w| w.of().set_bit() );
    tomu.TIMER1.cnt.write(|w| unsafe { w.cnt().bits(0u16) });

    efm32::NVIC::unpend(interrupt::TIMER1);
    tomu.NVIC.enable(interrupt::TIMER1);

    // start
    tomu.TIMER1.cmd.write(|w| w.start().set_bit());

    let timer1 = tomu.TIMER1;
    intr::free(|lock| ITIMER1.borrow(lock).replace(Some(timer1)));

    loop {
        asm::wfi();
    }

}

#[interrupt]
fn TIMER1() {
    intr::free(|lock| {
        if let (&mut Some(ref mut greenled), &mut Some(ref mut timer1)) = (
            GREENLED.borrow(lock).borrow_mut().deref_mut(),
            ITIMER1.borrow(lock).borrow_mut().deref_mut(),
        ) {
            timer1.cmd.write(|w| w.stop().set_bit());
            timer1.ifc.write(|w| w.of().set_bit());

            greenled.toggle();

            timer1.cnt.write(|w| unsafe { w.cnt().bits(0u16) });
            timer1.cmd.write(|w| w.start().set_bit());
        }
    });
}
```
</details>

I'll look some more into this.

**Comment by lucab** [comment]
I went back to my previous `capsense` experiment and I realized it is completely useless because:
 * I mixed some write vs modify calls (same as https://github.com/fudanchii/imtomu-rs/pull/33) ...
 * which resulted in a misconfigured ACMP, not really set in capsense mode ...
 * which resulted in some odd behavior (possibly my finger grounding positive and negative inputs together)

File: Cargo.toml [hunk 1]
```diff
@@ -33,6 +33,8 @@ compiletest_rs = "0.3.17"
 [features]
 toboot-custom-config = [ "tomu-macros" ]
 unproven = [ "embedded-hal/unproven", "efm32-hal/unproven" ]
+rt = [ ]
```
Comment (lucab) [inline Cargo.toml]: What is this feature doing?

File: examples/pac_capsense.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
+#![no_std]
```
Comment (lucab) [inline examples/pac_capsense.rs]: An high-level description here would be helpful, in order to cover what TIMER0, TIMER1 and PRS are doing.

File: examples/pac_capsense.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+#![no_main]
+
+extern crate panic_halt;
```
Comment (lucab) [inline examples/pac_capsense.rs]: `use panic_halt as _;`

File: examples/pac_capsense.rs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    measure_start();
+}
+
+fn measure_start() {
+    intr::free(|lock| if let (&mut Some(ref mut timer0), &mut Some(ref mut timer1)) = (
+            _TIMER0.borrow(lock).borrow_mut().deref_mut(),
+            _TIMER1.borrow(lock).borrow_mut().deref_mut(),
+    ) {
+        timer0.cnt.write(|w| unsafe { w.cnt().bits(0u16) });
+        timer1.cnt.write(|w| unsafe { w.cnt().bits(0u16) });
```
Comment (lucab) [inline examples/pac_capsense.rs]: I think this register and the one above have a `.reset()` method.

File: examples/pac_capsense.rs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    ) {
+        timer0.cmd.write(|w| w.stop().set_bit());
+        timer1.cmd.write(|w| w.stop().set_bit());
+        timer1.ifc.write(|w| w.of().set_bit());
+        timer0.cnt.read().cnt().bits()
+    } else { 0 })
+}
+
+fn clock_setup(tomu: &Tomu) {
+    tomu.CMU.hfperclken0.write(|w| w
```
Comment (lucab) [inline examples/pac_capsense.rs]: s/write/modify/ ?

File: examples/pac_capsense.rs [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    }
+}
+
+fn timer_setup(tomu: &Tomu) {
+    tomu.TIMER0.ctrl.write(|w| w
+        .presc().div1024()
+        .clksel().cc1()
+    );
+
+    tomu.TIMER0.top.write(|w| unsafe { w.bits(0xffffu32) });
```
Comment (lucab) [inline examples/pac_capsense.rs]: `.reset()` ?

File: examples/pac_capsense.rs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+            .sourcesel().acmp0()
+            .sigsel().bits(0u8)
+    });
+
+    tomu.TIMER1.ctrl.write(|w| w.presc().div1024());
+
+    // scan time 100ms
+    tomu.TIMER1.top.write(|w| unsafe { w.top().bits(2051u16) });
+    tomu.TIMER1.ien.write(|w| w.of().set_bit() );
+    tomu.TIMER1.cnt.write(|w| unsafe { w.cnt().bits(0u16) });
```
Comment (lucab) [inline examples/pac_capsense.rs]: `.reset()` ?

Comment (fudanchii) [reply examples/pac_capsense.rs]: I'll just remove this from here, covered by `measure_start`

File: examples/pac_capsense.rs (hunk 6 above)
Comment (fudanchii) [reply examples/pac_capsense.rs]: ah, yup the reset value is `0xffff` should use that instead :+1:

File: Cargo.toml (hunk 1 above)
Comment (fudanchii) [reply Cargo.toml]: we have guard when reimporting `efm32::interrupt` as `tomu::interrupt` this is to enable that.
I rearranged the feature in Cargo.toml, hope it makes more sense now

**Comment by lucab** [comment]
@fudanchii I actually figured out that `.ch1` works too.

The trick is that the bootloader uses that pin (`PC1`) in output mode in order to detect the "do not autorun, start Toboot" shorting manoeuvre. As such, before being able to use it we must put it back into analog input mode with a `gpio.pc1.into_input()`.

**Comment by fudanchii** [comment]
@lucab nice one! Sorry I haven't had the chance to work on this in the past weeks. Will take another look on it. Thank you~!

**Pull request merged**